mithril-client-cli/src/commands/tools/snapshot_converter.rs

```diff
@@ -0,0 +1,528 @@
+// TODO:
+// - Add logs.
+// - Remove the temporary directory in all cases (error or success).
+use std::path::Path;
+use std::process::Command;
+use std::{env, fmt, path::PathBuf};
+
+use anyhow::{anyhow, Context};
+use clap::{Parser, ValueEnum};
+
+use mithril_client::MithrilResult;
+
+use super::archive_unpacker_selector::ArchiveUnpackerSelector;
+use super::github_api_client::{GitHubApiClient, ReqwestGitHubApiClient};
+use super::http_downloader::{HttpDownloader, ReqwestHttpDownloader};
+
+const GITHUB_ORGANIZATION: &str = "IntersectMBO";
+const GITHUB_REPOSITORY: &str = "cardano-node";
+
+const LATEST_DISTRIBUTION_TAG: &str = "latest";
+const PRERELEASE_DISTRIBUTION_TAG: &str = "prerelease";
+
+const CARDANO_DISTRIBUTION_TEMP_DIR: &str = "cardano-node-distribution-tmp";
+
+const SNAPSHOT_CONVERTER_BIN_DIR: &str = "bin";
+const SNAPSHOT_CONVERTER_BIN_NAME_UNIX: &str = "snapshot-converter";
+const SNAPSHOT_CONVERTER_BIN_NAME_WINDOWS: &str = "snapshot-converter.exe";
+const SNAPSHOT_CONVERTER_CONFIG_DIR: &str = "share";
+const SNAPSHOT_CONVERTER_CONFIG_FILE: &str = "config.json";
+
+const LEDGER_DIR: &str = "ledger";
+
+#[derive(Debug, Clone, ValueEnum)]
+enum UTxOHDFlavor {
+    #[clap(name = "Legacy")]
+    Legacy,
+    #[clap(name = "LMDB")]
+    Lmdb,
+}
+
+impl fmt::Display for UTxOHDFlavor {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Legacy => write!(f, "Legacy"),
+            Self::Lmdb => write!(f, "LMDB"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, ValueEnum)]
+enum CardanoNetwork {
+    Preview,
+    Preprod,
+    Mainnet,
+}
+
+impl fmt::Display for CardanoNetwork {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Preview => write!(f, "preview"),
+            Self::Preprod => write!(f, "preprod"),
+            Self::Mainnet => write!(f, "mainnet"),
+        }
+    }
+}
+
+/// Clap command to convert a restored `InMemory` Mithril snapshot to another flavor.
+#[derive(Parser, Debug, Clone)]
+pub struct SnapshotConverterCommand {
+    /// Path to the Cardano node database directory.
+    #[clap(long)]
+    db_directory: PathBuf,
+
+    /// Cardano node version of the Mithril signed snapshot.
+    ///
+    /// `latest` and `prerelease` are also supported to download the latest or preprelease distribution.
+    #[clap(long)]
+    cardano_node_version: String,
+
+    /// Cardano network.
+    #[clap(long)]
+    cardano_network: CardanoNetwork,
+
+    /// UTxO-HD flavor to convert the ledger snapshot to.
+    #[clap(long)]
+    utxo_hd_flavor: UTxOHDFlavor,
+}
+
+impl SnapshotConverterCommand {
+    /// Main command execution
+    pub async fn execute(&self) -> MithrilResult<()> {
+        let distribution_temp_dir = self.db_directory.join(CARDANO_DISTRIBUTION_TEMP_DIR);
+        std::fs::create_dir(distribution_temp_dir.clone()).with_context(|| {
```

Comment:
Most `fs` methods takes `AsRef<Path>` as input which accept both moving the provided value or passing it by reference.
```suggestion
        std::fs::create_dir(&distribution_temp_dir).with_context(|| {
```